### PR TITLE
feat(cb2-8705): type approval number validation

### PR DESCRIPTION
**Pull request by Daniel-Searle**
## Type Approval Number validation

Type approval validators and formatting
[CB2-8705](https://dvsa.atlassian.net/browse/CB2-8705)

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/forms/components/approval-type/approval-type.component.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  @Input() isEditing = false;
+  @Input() approvalType?: string;
+
+  public form!: CustomFormGroup;
+
+  private techRecord_approvalTypeNumber1_: BehaviorSubject<string | undefined> = new BehaviorSubject<string | undefined>(undefined);
+  private techRecord_approvalTypeNumber2_: BehaviorSubject<string | undefined> = new BehaviorSubject<string | undefined>(undefined);
+  private techRecord_approvalTypeNumber3_: BehaviorSubject<string | undefined> = new BehaviorSubject<string | undefined>(undefined);
+  private techRecord_approvalTypeNumber4_: BehaviorSubject<string | undefined> = new BehaviorSubject<string | undefined>(undefined);
+  private techRecord_approvalTypeNumber1$: Observable<string | undefined>;
```
Comment (naathanbrown) [inline src/app/forms/components/approval-type/approval-type.component.ts:37]: I'd love a different variable name but I'm not sure what? This kind of implies that there are 4 numbers, when really it's 4 parts of 1 number

File: src/app/forms/components/switchable-input/switchable-input.component.ts [hunk 2]
```diff
@@ -25,6 +25,8 @@ export class SwitchableInputComponent implements OnInit {
   @Input() propOptions$?: Observable<MultiOptions>;
   @Input() hint?: string;
```
Comment (naathanbrown) [inline src/app/forms/components/switchable-input/switchable-input.component.ts:27]: This line isn't needed, tuck it up with the other optionals

File: src/app/forms/custom-sections/approval-type/approval-type.component.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  get isPsv(): boolean {
+    return this.techRecord.techRecord_vehicleType === VehicleTypes.PSV;
+  }
+
+  protected readonly TechRecord = TechRecord;
+  protected readonly getOptionsFromEnum = getOptionsFromEnum;
+  protected readonly approvalType = approvalType;
+  protected readonly console = console;
```
Comment (naathanbrown) [inline src/app/forms/custom-sections/approval-type/approval-type.component.ts:88]: Console? Do we need these how they are? I genuinely don't know the answer there

File: src/app/forms/templates/psv/psv-approval-type.template.ts [hunk 4]
```diff
@@ -57,14 +117,44 @@ export const PsvTypeApprovalTemplate: FormNode = {
...
+      ]
     },
     {
       name: 'techRecord_variantVersionNumber',
       label: 'Variant version number',
       type: FormNodeTypes.CONTROL,
       width: FormNodeWidth.XL,
-      validators: [{ name: ValidatorNames.MaxLength, args: 35 }]
+      validators: [
+        { name: ValidatorNames.RequiredIfEquals, args: { sibling: 'techRecord_approvalType', value: 'NTA' } },
```
Comment (naathanbrown) [inline src/app/forms/templates/psv/psv-approval-type.template.ts:143]: I wonder if we could make a 'requiredIfEquals' that accepts many values? Would make this really nice

**Comment by naathanbrown** [comment]
Also noticed that the IVA based selections aren't rendering out an input box.

If possible could the approval type box also scale to show all the text? Get that isn't on the ticket but nice thing to fix

File: src/app/forms/components/approval-type/approval-type.component.ts (hunk 1 above)
Comment (andrewf-bjss) [reply src/app/forms/components/approval-type/approval-type.component.ts:37]: Agreed, how about `techRecord_approvalTypeNumberPartx`

**Comment by naathanbrown** [comment]
Should I not get an error message if I miss out boxes? e.g. 
image

Also, I think if I change approval type the data should be wiped? Rather than formatting itself awkwardly

File: src/app/forms/components/approval-type/approval-type.component.ts (hunk 1 above)
Comment (naathanbrown) [reply src/app/forms/components/approval-type/approval-type.component.ts:37]: Not sure we even need the techRecord? These values are not on a techRecord

Comment (Daniel-Searle) [reply src/app/forms/components/approval-type/approval-type.component.ts:37]: approvalTypeNumberPartx sounds good to me @naathanbrown  @andrewf-bjss

File: src/app/forms/components/approval-type/approval-type.component.ts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  styleUrls: ['./approval-type.component.scss'],
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: ApprovalTypeInput,
+      multi: true
+    }
+  ]
+})
+export class ApprovalTypeInput extends BaseControlComponent implements OnInit, OnDestroy, AfterContentInit {
```
[the remark this answers is not in the document]
Comment (Daniel-Searle) [reply src/app/forms/components/approval-type/approval-type.component.ts:27]: I need to change it to ApprovalTypeInputComponent right?

File: src/app/forms/components/approval-type/approval-type.component.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+          };
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  ngOnChanges(): void {
+    console.log(this.approvalTypeChange);
```
Comment (tomcrawleyy) [inline src/app/forms/components/approval-type/approval-type.component.ts:157]: Leftover console log?

File: src/app/forms/components/approval-type/approval-type.component.ts [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            ]
+          };
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  ngOnChanges(): void {
```
Comment (naathanbrown) [inline src/app/forms/components/approval-type/approval-type.component.ts:156]: Need to fix this, basically just need to update what it implements

File: src/app/forms/templates/psv/psv-approval-type.template.ts [hunk 8]
```diff
@@ -21,28 +21,88 @@ export const PsvTypeApprovalTemplate: FormNode = {
       label: 'Approval type number',
       type: FormNodeTypes.CONTROL,
       width: FormNodeWidth.XL,
-      validators: [{ name: ValidatorNames.MaxLength, args: 25 }]
+      validators: [
+        { name: ValidatorNames.RequiredIfEquals, args: { sibling: 'techRecord_approvalType', value: 'NTA' } },
```
Comment (naathanbrown) [inline src/app/forms/templates/psv/psv-approval-type.template.ts:25]: apply new validator change here

**Review by naathanbrown — DISMISSED** [review]

File: src/app/forms/components/approval-type/approval-type.component.ts [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+  constructor(injector: Injector, changeDetectorRef: ChangeDetectorRef, public globalErrorService: GlobalErrorService) {
+    super(injector, changeDetectorRef);
+    this.approvalTypeNumber1$ = this.approvalTypeNumber1_.asObservable();
+    this.approvalTypeNumber2$ = this.approvalTypeNumber2_.asObservable();
+    this.approvalTypeNumber3$ = this.approvalTypeNumber3_.asObservable();
+    this.approvalTypeNumber4$ = this.approvalTypeNumber4_.asObservable();
+    this.globalErrorService.errors$.subscribe((globalErrors: any) => {
+      if (globalErrors.length) {
+        this.formSubmitted = true;
+        console.log(globalErrors);
```
Comment (tomcrawleyy) [inline src/app/forms/components/approval-type/approval-type.component.ts:76]: leftover console log

**Review by tomcrawleyy — APPROVED** [review]

**Review by naathanbrown — APPROVED** [review]

**Review by tomcrawleyy — APPROVED** [review]

**Pull request merged**